[jit] Add dictionary as output of tracer

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31860 [jit] Add dictionary as output of tracer**
* #36609 [jit] add comment for Dict iterationOrder

Summary:

This add dictionary as a supported output of tracer.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20638438](https://our.internmc.facebook.com/intern/diff/D20638438)

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 41fa3f18:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=41fa3f18248362e825b0fccee664d1c193f98f5b).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4133668) pytorch_linux_xenial_py3_6_gcc5_4_ge_config_simple_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=80379555) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=80379555))
<details>
<summary>
<code>Jan 04 19:17:54 RuntimeError: test_jit_simple failed!</code>
</summary>

```
Jan 04 19:17:54 Ran 2313 tests in 74.965s 
Jan 04 19:17:54  
Jan 04 19:17:54 FAILED (errors=1, skipped=82, expected failures=1) 
Jan 04 19:17:54  
Jan 04 19:17:54 Generating XML reports... 
Jan 04 19:17:54 Traceback (most recent call last): 
Jan 04 19:17:54   File "test/run_test.py", line 456, in <module> 
Jan 04 19:17:54     main() 
Jan 04 19:17:54   File "test/run_test.py", line 449, in main 
Jan 04 19:17:54     raise RuntimeError(message) 
Jan 04 19:17:54 RuntimeError: test_jit_simple failed! 
Jan 04 19:17:54 + cleanup 
Jan 04 19:17:54 + retcode=1 
Jan 04 19:17:54 + set +x 
Jan 04 19:17:54 =================== sccache compilation log =================== 
Jan 04 19:17:54 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 04 19:17:54 Compile requests                 2 
Jan 04 19:17:54 Compile requests executed        0 
Jan 04 19:17:54 Cache hits                       0 
Jan 04 19:17:54 Cache misses                     0 
Jan 04 19:17:54 Cache timeouts                   0 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

wanchaol committed:
[jit] Add dictionary as output of tracer

bddppq committed:
Sort Tensor keys by their unique ids to make the traced graph deterministic

bddppq committed:
Merge branch 'gh/wanchaol/75/base' into gh/wanchaol/75/head

jamesr66a commented on test/test_jit.py:
```diff
@@ -4184,6 +4184,25 @@ def forward(self, x):
+
+        x = (torch.rand(3), torch.rand(3))
+        traced_dict_str_mod = torch.jit.trace(TraceDictStrTensor(), x)
+        print(traced_dict_str_mod.graph)
```
remove

jamesr66a commented on test/test_jit.py:
```diff
@@ -4184,6 +4184,25 @@ def forward(self, x):
+        self.assertEqual(traced_dict_str_mod(*x), {'a': x[0], 'b': x[1]})
+
+        traced_dict_tensor_mod = torch.jit.trace(TraceDictTensorTensor(), x)
+        print(traced_dict_tensor_mod.graph)
```
remove

jamesr66a reviewed (commented):
This seems like it would add very non-obvious behavior: A user could write a program like

```
import torch

def foo(x):
  rv = {}
  for i in range(x.size(0)):
    rv["{}".format(i)] = x[i]
  return rv

traced = torch.jit.trace(foo, (torch.rand(3, 4),))

# OK
print(traced(torch.rand(3, 4)))
# Not what the user intended!
print(traced(torch.rand(5, 6)))
```

```
{'2': tensor([0.2862, 0.7441, 0.5652, 0.4712]), '1': tensor([0.9654, 0.2331, 0.0763, 0.9142]), '0': tensor([0.5700, 0.5072, 0.1533, 0.9637])}
{'2': tensor([0.0182, 0.4806, 0.7015, 0.6835, 0.1624, 0.0149]), '1': tensor([0.6804, 0.2785, 0.9390, 0.7019, 0.3990, 0.0540]), '0': tensor([0.0310, 0.0017, 0.9871, 0.2879, 0.7661, 0.5934])}
```

Can you elaborate on the use case here? The only time this would make sense is if the structure of the dict is fixed, in which case you should probably use a NamedTuple or some other data structure

wanchaol commented:
> This seems like it would add very non-obvious behavior: A user could write a program like
> 
> ```
> import torch
> 
> def foo(x):
>   rv = {}
>   for i in range(x.size(0)):
>     rv["{}".format(i)] = x[i]
>   return rv
> 
> traced = torch.jit.trace(foo, (torch.rand(3, 4),))
> 
> # OK
> print(traced(torch.rand(3, 4)))
> # Not what the user intended!
> print(traced(torch.rand(5, 6)))
> ```
> 
> ```
> {'2': tensor([0.2862, 0.7441, 0.5652, 0.4712]), '1': tensor([0.9654, 0.2331, 0.0763, 0.9142]), '0': tensor([0.5700, 0.5072, 0.1533, 0.9637])}
> {'2': tensor([0.0182, 0.4806, 0.7015, 0.6835, 0.1624, 0.0149]), '1': tensor([0.6804, 0.2785, 0.9390, 0.7019, 0.3990, 0.0540]), '0': tensor([0.0310, 0.0017, 0.9871, 0.2879, 0.7661, 0.5934])}
> ```
> 
> Can you elaborate on the use case here? The only time this would make sense is if the structure of the dict is fixed, in which case you should probably use a NamedTuple or some other data structure

@jamesr66a The use case is that user would like a more general dictionary like output in the tracer, some user could not use named tuple because of several constraints, i.e. it restricts field names to valid python identifiers that do not begin with underscore. 

For the pattern that you identified, I agreed that's something that user won't expect, but given that this is a data dependent control flow, this falls into the same area that tracer won't be able to handle anyways and user should have aware that when they use tracing. Also, we already have List, Dict supported as tracer input, the same problem also arise there, so we should match the input/output types supported by tracer.

jamesr66a commented:
> some user could not use named tuple because of several constraints, i.e. it restricts field names to valid python identifiers that do not begin with underscore.

It sounds like we should address this rather than adding functionality to the tracer. What is causing the user to not be able to make their field names valid IDs?

wanchaol committed:
Update on "[jit] Add dictionary as output of tracer"


Summary:

This add dictionary as a supported output of tracer.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20638438](https://our.internmc.facebook.com/intern/diff/D20638438)

[ghstack-poisoned]

fritzo commented:
Hi @jamesr66a , one example where it is prohibitively difficult to use a `namedtuple` is in Pyro https://github.com/pyro-ppl/pyro/pull/2177 . Pyro's guides are `nn.Module`s that return a dict mapping sample site name to sampled value. Pyro accepts arbitrary strings as sample site names, as specified by Pyro users (not by me 😄).

One workaround @neerajprad and I have discussed is to use something a hex encoder to convert to a string that PyTorch accepts (i.e. a valid `namedtuple` key i.e. a python identifier that does not begin with underscore), the add a decoding layer to convert from encoded namedtuples to decoded dicts. However that decoding layer would also need to be jittable so that dict keys are interpretable from C++.

Very much looking forward to dictionary support! 🎉

bddppq commented:
Beside the issue of valid python identifier, there is another valid use case to use dict: Imagine there is a backbone module shared across multiple models, each model only adds head(s) after the backbone, so:
model1: backbone -> head1
model2: backbone -> head2 
...
within each model, the features are static, however across different models the features can be different. Since backbone is provided as a library to other model authors to write their own heads to build their models, it has no assumption on the full set of features across different models/heads, and so uses dict to pass data (`Dict[str, Tensor]`) between backbone and heads.

bddppq commented:
Also links #27743

jamesr66a commented:
Alright, the main gap in this implementation is that we need to require the user to certify "I know for a fact that this dict does not change schema, please record it structurally as you see it". Using NamedTuple is one way to do this, but I've heard several arguments about why this makes things difficult.

Let's try this: can we gate this return behavior behind a flag? i.e. If I do something like:

```
class Foo(torch.nn.Module):
  def forward(self, x):
    return {'a' : x, 'b' : x + 1}

traced = torch.jit.trace(Foo(), (torch.rand(3, 4),))
```
It would throw:
```
RuntimeError: Encountered a dictionary at the output of the traced callable! This is only valid if the structure of the dictionary (i.e. keys) does not change based on the module's inputs. If this is the case, pass allow_dict_output=True to trace() to allow this behavior.
```

Then if I do
```
traced = torch.jit.trace(Foo(), (torch.rand(3, 4),), allow_dict_output=True)
```

It should work as implemented here.

The reason for all this is we don't want incorrectly-recorded programs to silently have wrong behavior and potentially make it into a production setting and cause major issues

bddppq commented:
The surprising behavior demonstrated in https://github.com/pytorch/pytorch/pull/31860#pullrequestreview-381693140 is actually not caused by the limitation of static dict, but rather the size Tensor has been casted to Python int in the `range(x.size(0))` expression. With other container types tracing has similar behavior:

```python
import torch

def foo_list(x):
  rv = []
  for i in range(x.size(0)):
    rv.append(x[i])
  return rv

traced = torch.jit.trace(foo_list, (torch.randn(3, 1)))
print(traced(torch.randn(3, 1)))
print(traced(torch.randn(5, 1)))

def foo_tuple(x):
    return tuple(x[i] for i in range(x.size(0)))

traced = torch.jit.trace(foo_tuple, (torch.randn(3, 1)))
print(traced(torch.randn(3, 1)))
print(traced(torch.randn(5, 1)))
```
```
 TracerWarning: Converting a tensor to a Python index might cause the trace to be incorrect. We can't record the data flow of P
ython values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  for i in range(x.size(0)):
[tensor([-0.0323]), tensor([-2.0990]), tensor([0.3493])]
[tensor([-0.3254]), tensor([1.0486]), tensor([-0.7183])]
TracerWarning: Converting a tensor to a Python index might cause the trace to be incorrect. We can't record the data flow of
Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  return tuple(x[i] for i in range(x.size(0)))
(tensor([0.5241]), tensor([-0.1070]), tensor([-2.0607]))
(tensor([-1.0013]), tensor([1.5692]), tensor([-0.1939]))
```

bddppq commented:
So if we are going to add a flag to prevent this silently wrong behavior, maybe it should be guarding the casting (but it will be a breaking change, right now the tracer only gives warning).

wanchaol commented:
> So if we are going to add a flag to prevent this silently wrong behavior, maybe it should be guarding the casting (but it will be a breaking change, right now the tracer only gives warning).

@bddppq you are right, the case only happens in dynamic control flow, and casting currently gives a warning to the user that might get ignored by user who don't read the doc and understand the internals of tracer. The general issue might not only happening for tensor casting though, i.e. user could check `len(list)` or `len(dict)` inside the tracer in the control flow and do whatever they want, but a changed input might be wrong there too. 

Had a discussion with @jamesr66a and other folks today. The short term plan is to guard the risk behavior under a `strict` flag, and throw warnings/errors to user. Right now tensor casting will throw warning, others not. To ensure not BC breaking with existing user code too much, we should probably:

1. throw warning in tracer when discovering list (input/output) + dict(input), these are thing already supported but no warning yet. 
2. for dict output, we will throw error and explicitly guide the user to use `strict=False` flag.

wanchaol committed:
Update on "[jit] Add dictionary as output of tracer"


Summary:

This add dictionary as a supported output of tracer.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20638438](https://our.internmc.facebook.com/intern/diff/D20638438)

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] Add dictionary as output of tracer"


Summary:

This add dictionary as a supported output of tracer.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20638438](https://our.internmc.facebook.com/intern/diff/D20638438)

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] Add dictionary as output of tracer"


Summary:

This add dictionary as a supported output of tracer.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20638438](https://our.internmc.facebook.com/intern/diff/D20638438)

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] Add dictionary as output of tracer"


Summary:

This add dictionary as a supported output of tracer.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20638438](https://our.internmc.facebook.com/intern/diff/D20638438)

[ghstack-poisoned]

jamesr66a commented on torch/csrc/jit/frontend/tracer.cpp:
```diff
@@ -242,9 +242,48 @@ Value* TracingState::getOutput(const IValue& iv, size_t i) {
+    TypePtr key_type = dict.keyType();
+    TypePtr value_type = dict.valueType();
+
+    if (!(key_type->cast<StringType>() || key_type->cast<TensorType>()) ||
```
This is kinda hard to read in the !(A || B) || !C form. Can we do something like:
```
bool key_type_valid = key_type->cast<StringType>() || key_type->cast<TensorType>();
bool value_type_valid = value_type->cast<TensorType>();

if (!key_type_valid || !value_type_valid) {
  <...>
```

jamesr66a commented on torch/csrc/jit/frontend/tracer.cpp:
```diff
@@ -242,9 +242,48 @@ Value* TracingState::getOutput(const IValue& iv, size_t i) {
+         << "or dict[Tensor, Tensor] can be a dictionary output of a traced function";
+      throw std::runtime_error(os.str());
+    }
+    const auto order = iterationOrder(dict);
```
Not an issue with your PR, but I looked at the declaration of `iterationOrder` and I'm not sure what it does :p. It could use some comments or a rename to indicate its function

jamesr66a replied:
Can you (or someone) add that as a comment?

wanchaol replied:
did this in https://github.com/pytorch/pytorch/pull/36609

jamesr66a commented on torch/csrc/jit/frontend/tracer.cpp:
```diff
@@ -242,9 +242,48 @@ Value* TracingState::getOutput(const IValue& iv, size_t i) {
+      std::iota(perm.begin(), perm.end(), 0);
+      std::sort(perm.begin(), perm.end(), [&](int64_t i, int64_t j) {
+        return keys[i]->unique() < keys[j]->unique();
+      });
```
Can you comment on the stability of these unique IDs? Doesn't Value::unique() potentially depend on the order of compilation of values?

jamesr66a reviewed (commented):
Cool, I had a question about how we're doing ordering for Tensor-keyed dictionaries. I want to make sure that's safe and stable across compilation/machines etc

wanchaol committed:
Update on "[jit] Add dictionary as output of tracer"


Summary:

This add dictionary as a supported output of tracer.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20638438](https://our.internmc.facebook.com/intern/diff/D20638438)

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] Add dictionary as output of tracer"


Summary:

This add dictionary as a supported output of tracer.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20638438](https://our.internmc.facebook.com/intern/diff/D20638438)

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] Add dictionary as output of tracer"


Summary:

This add dictionary as a supported output of tracer.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20638438](https://our.internmc.facebook.com/intern/diff/D20638438)

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] Add dictionary as output of tracer"


Summary:

This add dictionary as a supported output of tracer.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20638438](https://our.internmc.facebook.com/intern/diff/D20638438)

[ghstack-poisoned]

wanchaol commented:
something weird happens with ghstack on this PR, replace this with https://github.com/pytorch/pytorch/pull/36696
